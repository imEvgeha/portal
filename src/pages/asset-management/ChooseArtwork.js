import React, {useState} from 'react';
import ArtworkActions from './artwork-actions/ArtworkActions';
import ArtworkItem from './artwork-item/ArtworkItem';
import {LIST} from './constants';
import './ChooseArtwork.scss';

const ChooseArtwork = () => {
    const [selectedItems, setSelectedItems] = useState([]);
    const [totalItems, setTotalItems] = useState(LIST);

    const itemClick = id => {
        if (selectedItems.includes(id)) {
            setSelectedItems(selectedItems.filter(value => value !== id));
        } else {
            setSelectedItems([...selectedItems, id]);
        }
    };

    const setItems = () => {
        selectedItems.length ? setSelectedItems([]) : setSelectedItems(totalItems.map(item => item.id));
    };

    return (
        <div className="choose-artwork">
            <div className="choose-artwork__header">
                <span>Title: </span>
                <span>Star Wars (1977)</span>
            </div>
            <ArtworkActions selectedItems={selectedItems} totalItems={totalItems} setItems={setItems} />
            <div className="choose-artwork__list">
                {LIST.map(item => (
                    <ArtworkItem
                        key={item.id}
                        item={item}
                        onClick={() => itemClick(item.id)}
                        isSelected={selectedItems.includes(item.id)}
                    />
                ))}
            </div>
        </div>
    );
};

ChooseArtwork.propTypes = {};

ChooseArtwork.defaultProps = {};

export default ChooseArtwork;
