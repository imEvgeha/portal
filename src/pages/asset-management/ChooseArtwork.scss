@import '../../styles/atlaskit/colors';

.choose-artwork {
    background: map-get($neutrals, N0);
    &__header {
        padding: 10px 24px;
        background: map-get($neutrals, N20);
        color: map-get($neutrals, N800);
        font-size: 20px;
        span {
            margin-right: 5px;
        }
    }
    &__list {
        display: flex;
        flex-wrap: wrap;
        padding: 24px;
    }
}
