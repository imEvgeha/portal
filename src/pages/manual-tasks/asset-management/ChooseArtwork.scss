@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/colors';

.choose-artwork {
    background: map-get($neutrals, N0);
    display: flex;
    flex-flow: column;
    height: 100%;
    &__header {
        flex: 0 1 auto;
        padding: 10px 24px;
        background: map-get($neutrals, N20);
        color: map-get($neutrals, N800);
        font-size: 20px;
        span {
            margin-right: 5px;
        }
    }
    &__list {
        display: flex;
        flex-wrap: wrap;
        padding: 24px;
    }

    .artwork-items {
        flex: 1 1 auto;
    }

    .scroll-item {
        display: inline-block;
        height: 200px;
    }

    .loading-item {
        height: 200px;
        display: inline-block;
        width: 300px;
        .skeleton {
            width: 100% !important;
            height: 100% !important;
        }
    }

    .artwork-actions {
        flex: 0 1 40px;
        display: flex;
        align-items: center;
        margin: 0 20px;
        height: 54px;
        padding: 10px 0;
        border-bottom: 2px solid map-get($neutrals, N40);
        label {
            margin: 0;
        }
    }
}
