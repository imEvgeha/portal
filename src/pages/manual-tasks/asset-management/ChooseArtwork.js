import React, {useEffect, useState, useCallback, useContext} from 'react';
import PropTypes from 'prop-types';
import {Button} from '@portal/portal-components';
import {NexusModalContext} from '@vubiquity-nexus/portal-ui/lib/elements/nexus-modal/NexusModal';
import {URL as VuURL} from '@vubiquity-nexus/portal-utils/lib/Common';
import DOP from '@vubiquity-nexus/portal-utils/lib/DOP';
import {Skeleton} from 'primereact/skeleton';
import {VirtualScroller} from 'primereact/virtualscroller';
import {connect, useDispatch} from 'react-redux';
import Loading from '../../static/Loading';
import './ChooseArtwork.scss';
import {fetchPosters, fetchAsset, removeMediaIngest} from './assetManagementReducer';
import {posterListSelector, assetDetailsSelector, mediaIngestsSelector} from './assetManagementSelectors';
import {fetchPoster, loginAssets} from './assetManagementService';
import UploadArtworkForm from './components/UploadArtworkForm';
import ArtworkItem from './components/artwork-item/ArtworkItem';

const UPLOAD_ARTWORK_TITLE = 'Upload Artwork';
const DOP_POP_UP_TITLE = 'Choose Artwork';
const DOP_POP_UP_MESSAGE = 'Please select at least one image';
const IMG_WIDTH = 300;
const IMG_HEIGHT = 200;

const ChooseArtwork = ({fetchResourcePosters, posterList, fetchAsset, asset, jobIds}) => {
    const dispatch = useDispatch();
    const [selectedArtwork, setSelectedArtwork] = useState();
    const [posters, setPosters] = useState([]);
    const [lazyLoading, setLazyLoading] = useState(false);
    const [itemSize] = useState(Math.trunc((window.screen.width - 50) / IMG_WIDTH));

    const {openModal, closeModal} = useContext(NexusModalContext);
    const sourceMediaAssetID = VuURL.getParamIfExists('sourceMediaAssetID', '');
    const artworkAssetID = VuURL.getParamIfExists('artworkAssetID', '');

    const openDOPPopUp = useCallback(() => {
        const handlePopUpClick = () => {
            DOP.sendInfoToDOP(1, null);
            closeModal();
        };
        const actions = [
            {
                text: 'OK',
                onClick: handlePopUpClick,
                appearance: 'primary',
            },
        ];
        openModal(DOP_POP_UP_MESSAGE, {title: DOP_POP_UP_TITLE, width: 'medium', actions});
    }, []);

    const openUploadModal = () => {
        openModal(<UploadArtworkForm closeModal={closeModal} tenantId={asset.tenantId} asset={asset} />, {
            title: UPLOAD_ARTWORK_TITLE,
            width: 'medium',
            shouldCloseOnOverlayClick: false,
        });
    };

    useEffect(() => {
        setPosters(Array.from({length: posterList.length}));
        setLazyLoading(false);
    }, [posterList]);

    useEffect(() => {
        fetchAsset(artworkAssetID);
        loginAssets().then(() => fetchResourcePosters(sourceMediaAssetID));
    }, []);

    useEffect(() => {
        DOP.setDOPMessageCallback(selectedArtwork || jobIds.length !== 0 ? null : () => openDOPPopUp());
    }, [selectedArtwork, jobIds, openDOPPopUp]);

    const artworkClick = (id, uri) => {
        setSelectedArtwork(id);
        DOP.setErrorsCount(0);

        DOP.setData({
            chooseArtwork: {
                sourceMediaAssetID,
                selectedArtworkUri: uri,
            },
        });
        dispatch(removeMediaIngest());
    };

    const basicItemTemplate = item => {
        const timing = item?.url?.split('/')?.at(-1);
        return (
            <div className="nexus-c-scroll-item">
                <ArtworkItem
                    key={timing}
                    poster={item?.img}
                    timing={timing}
                    onClick={() => artworkClick(timing, item?.url)}
                    isSelected={selectedArtwork === timing}
                />
            </div>
        );
    };

    const onLazyLoad = event => {
        const {first, last} = event;
        const positionsToCheck = posters.slice(first, last);
        if (posters.includes(undefined) && positionsToCheck.includes(undefined) && !lazyLoading) {
            setLazyLoading(true);

            const notDefinedItems = [];
            positionsToCheck.forEach((item, index) => !item && notDefinedItems.push(index));
            const firstItem = notDefinedItems[0] + first;
            const lastItem = first + notDefinedItems.at(-1) + 1;
            const postersInView = posterList.slice(firstItem, lastItem);
            const postersToFetch = postersInView.map(url => fetchPoster(url));
            const lazyItems = [...posters];

            Promise.all(postersToFetch).then(res => {
                let counter = 0;
                for (let i = firstItem; i < lastItem; i++) {
                    lazyItems[i] = {img: URL.createObjectURL(res[counter]), url: postersInView[counter]};
                    counter++;
                }
                setLazyLoading(false);
                setPosters(lazyItems);
            });
        }
    };

    const basicLoadingTemplate = () => {
        return (
            <div className="nexus-c-loading-item px-2">
                <Skeleton width="100%" height="100%" />
            </div>
        );
    };

    return (
        <div className="choose-artwork">
            <div className="choose-artwork__header">
                <span>Title: </span>
                <span>{VuURL.getParamIfExists('title', '')}</span>
            </div>
            <div className="artwork-actions">
                {asset ? (
                    <Button className="p-button-outlined" label="Upload" onClick={openUploadModal} />
                ) : (
                    <Loading />
                )}
            </div>
            <div className="nexus-c-artwork-items">
                <VirtualScroller
                    items={posters}
                    itemSize={Math.trunc(IMG_HEIGHT / itemSize)}
                    itemTemplate={basicItemTemplate}
                    scrollHeight="100%"
                    lazy
                    onLazyLoad={onLazyLoad}
                    showLoader
                    loading={lazyLoading}
                    loadingTemplate={basicLoadingTemplate}
                />
            </div>
        </div>
    );
};

ChooseArtwork.propTypes = {
    fetchResourcePosters: PropTypes.func.isRequired,
    fetchAsset: PropTypes.func.isRequired,
    asset: PropTypes.object,
    posterList: PropTypes.array,
    jobIds: PropTypes.array,
};

ChooseArtwork.defaultProps = {
    posterList: [],
    asset: null,
    jobIds: [],
};

const mapStateToProps = state => ({
    posterList: posterListSelector(state),
    asset: assetDetailsSelector(state),
    jobIds: mediaIngestsSelector(state),
});

const mapDispatchToProps = dispatch => ({
    fetchResourcePosters: sourceMediaAssetID => dispatch(fetchPosters(sourceMediaAssetID)),
    fetchAsset: sourceMediaAssetID => dispatch(fetchAsset(sourceMediaAssetID)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ChooseArtwork);
