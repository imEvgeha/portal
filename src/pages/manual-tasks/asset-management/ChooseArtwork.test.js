import React from 'react';
import {shallow} from 'enzyme';
import configureStore from 'redux-mock-store';
import ChooseArtwork from './ChooseArtwork';

describe('ChooseArtwork', () => {
    let wrapper = null;
    const mockStore = configureStore();
    const store = mockStore({
        manualTasks: {
            assets: {
                posterList: ['abc'],
            },
        },
    });

    beforeAll(() => {
        wrapper = shallow(<ChooseArtwork store={store} />)
            .find('ChooseArtwork')
            .shallow();
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render 1 item', () => {
        expect(wrapper.find('VirtualScroller').length).toEqual(1);
    });
});
