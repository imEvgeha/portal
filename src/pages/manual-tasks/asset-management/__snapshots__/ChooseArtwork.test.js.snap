// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChooseArtwork should match snapshot 1`] = `
<div
  className="choose-artwork"
>
  <div
    className="choose-artwork__header"
  >
    <span>
      Title: 
    </span>
    <span />
  </div>
  <ArtworkActions
    posterList={
      Array [
        "abc",
      ]
    }
    selectedItems={0}
    setSelectedItems={[Function]}
  />
  <div
    className="choose-artwork__list"
  >
    <ArtworkItem
      isSelected={false}
      key="abc"
      onClick={[Function]}
      poster="abc"
      timing="abc"
    />
  </div>
</div>
`;
